src/TurbulenceClosures/turbulence_closure_implementations/CATKEVerticalDiffusivities/time_step_turbulent_kinetic_energy.jl

```diff
@@ -0,0 +1,206 @@
+    end
+
+    FT = eltype(grid)
+
+    for m = 1:M # substep
+        if m == 1 && M != 1
+            χ = convert(FT, -0.5) # Euler step for the first substep
+        else
+            χ = model.timestepper.χ
+        end
+
+        # Compute the linear implicit component of the RHS (diffusivities, L)
+        # and step forward
+        launch!(arch, grid, :xyz,
+                substep_turbulent_kinetic_energy!,
+                κe, Le, grid, closure,
+                model.velocities, previous_velocities, # try this soon: model.velocities, model.velocities,
+                model.tracers, model.buoyancy, diffusivity_fields,
+                Δτ, χ, Gⁿe, G⁻e)
+
+        # Good idea?
+        # previous_time = model.clock.time - Δt
+        # previous_iteration = model.clock.iteration - 1
+        # current_time = previous_time + m * Δτ
+        # previous_clock = (; time=current_time, iteration=previous_iteration)
+
+        implicit_step!(e, implicit_solver, closure,
+                       model.diffusivity_fields, Val(tracer_index),
+                       model.clock, Δτ)
+    end
+
+    return nothing
+end
+
+@kernel function substep_turbulent_kinetic_energy!(κe, Le, grid, closure,
+                                                   next_velocities, previous_velocities,
+                                                   tracers, buoyancy, diffusivities,
+                                                   Δτ, χ, slow_Gⁿe, G⁻e)
+
+    i, j, k = @index(Global, NTuple)
+
+    Jᵇ = diffusivities.Jᵇ
+    e = tracers.e
+    closure_ij = getclosure(i, j, closure)
+
+    # Compute TKE diffusivity.
+    κe★ = κeᶜᶜᶠ(i, j, k, grid, closure_ij, next_velocities, tracers, buoyancy, Jᵇ)
+    κe★ = mask_diffusivity(i, j, k, grid, κe★)
+    @inbounds κe[i, j, k] = κe★
+
+    # Compute additional diagonal component of the linear TKE operator
+    wb = explicit_buoyancy_flux(i, j, k, grid, closure_ij, next_velocities, tracers, buoyancy, diffusivities)
+    wb⁻ = min(zero(grid), wb)
+    wb⁺ = max(zero(grid), wb)
+
+    eⁱʲᵏ = @inbounds e[i, j, k]
+    eᵐⁱⁿ = closure_ij.minimum_turbulent_kinetic_energy
+    wb⁻_e = wb⁻ / eⁱʲᵏ * (eⁱʲᵏ > eᵐⁱⁿ)
+
+    # Treat the divergence of TKE flux at solid bottoms implicitly.
+    # This will damp TKE near boundaries. The bottom-localized TKE flux may be written
+    #
+    #       ∂t e = - δ(z + h) ∇ ⋅ Jᵉ + ⋯
+    #       ∂t e = + δ(z + h) Jᵉ / Δz + ⋯
+    #
+    # where δ(z + h) is a δ-function that is 0 everywhere except adjacent to the bottom boundary
+    # at $z = -h$ and Δz is the grid spacing at the bottom
+    #
+    # Thus if
+    #
+    #       Jᵉ ≡ - Cᵂϵ * √e³
+    #          = - (Cᵂϵ * √e) e
+    #
+    # Then the contribution of Jᵉ to the implicit flux is
+    #
+    #       Lᵂ = - Cᵂϵ * √e / Δz.
+    #
+    on_bottom = !inactive_cell(i, j, k, grid) & inactive_cell(i, j, k-1, grid)
+    Δz = Δzᶜᶜᶜ(i, j, k, grid)
+    Cᵂϵ = closure_ij.turbulent_kinetic_energy_equation.Cᵂϵ
+    e⁺ = clip(eⁱʲᵏ)
```

Comment:
what does `clip` do? where is `clip` defined?

Reply:
in another file in this module

Reply:
`clip` limits `e` to positive values, eg `max(zero(e), e)` to be type stable. Maybe we don't need it...

Reply:
It’s ok. But where does it come from? It’s not in Julia base..

Reply:
ok!

Reply:
I believe it's defined within this same module! I put a comment in

Reply:
all good, I found it... no need to comment